### PR TITLE
UAR-90: Created update journey interrupt card page

### PR DESCRIPTION
**Pull request by azafarCH**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-90

### Change description
- Created update journey interrupt card page

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/update/update.interrupt.card.controller.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import { logger } from "../../utils/logger";
+import * as config from "../../config";
+
+export const get = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `${req.method} ${req.route.path}`);
+
+    return res.render(config.UPDATE_INTERRUPT_CARD_PAGE, {
+      backLinkUrl: "",
```
Comment (AndreKramer7654) [inline src/controllers/update/update.interrupt.card.controller.ts:11]: Is the back link the use paper page? Assume it's interrupting the forward from that ...

File: src/routes/index.ts [hunk 2]
```diff
@@ -238,6 +239,11 @@ router.route(config.UPDATE_USE_PAPER_URL)
   .all(authentication)
   .get(updateUsePaper.get);
 
+router.route(config.UPDATE_INTERRUPT_CARD_URL)
+  .all(authentication)
+  .get(updateInterruptCard.get);
+// .post(updateInterruptCard.post);
```
Comment (AndreKramer7654) [inline src/routes/index.ts:245]: remove commented out code

File: test/controllers/update/update.interrupt.card.controller.spec.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import { UPDATE_INTERRUPT_CARD_URL, UPDATE_INTERRUPT_CARD_PAGE } from "../../../src/config";
+import { INTERRUPT_CARD_PAGE_TITLE } from "../../__mocks__/text.mock";
+
+import {
+  ANY_MESSAGE_ERROR,
+  SERVICE_UNAVAILABLE,
+} from "../../__mocks__/text.mock";
+
+import { authentication } from "../../../src/middleware/authentication.middleware";
+// import { isSecureRegister } from "../../../src/middleware/navigation/is.secure.register.middleware";
```
Comment (AndreKramer7654) [inline test/controllers/update/update.interrupt.card.controller.spec.ts:21]: remove commented out code

**Comment by AndreKramer7654** [comment]
Do we need navigation control here? Don't think we are recording anything in the session so nothing to check?

File: src/controllers/update/update.interrupt.card.controller.ts (hunk 1 above)
Comment (azafarCH) [reply src/controllers/update/update.interrupt.card.controller.ts]: Added the backlink now for the secure update filter page.

File: views/update/update-interrupt-card.html [hunk 4]
```diff
@@ -0,0 +1,42 @@
+{% extends "layout.html" %}
+
+{% set title = "Before you start" %}
```
Comment (sjahir) [inline views/update/update-interrupt-card.html:3]: Need to add `(NOT LIVE)` in the title: `Before you start (NOT LIVE)`

File: test/controllers/update/secure.update.filter.controller.spec.ts [hunk 5]
```diff
@@ -82,12 +82,12 @@ describe("SECURE UPDATE FILTER controller", () => {
   });
 
   describe("POST tests", () => {
-    test("redirect to overseas entity query page if user selects no", async () => {
+    test("redirect to before you start page if user selects no", async () => {
```
Comment (sjahir) [inline test/controllers/update/secure.update.filter.controller.spec.ts:85]: Consider using the name of the html file so: `redirect to update interrupt card page...`

File: test/controllers/update/update.interrupt.card.controller.spec.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+describe("UPDATE INTERRUPT CARD controller", () => {
+  describe("GET tests", () => {
+    test(`renders the ${UPDATE_INTERRUPT_CARD_PAGE} page`, async () => {
+      const resp = await request(app).get(UPDATE_INTERRUPT_CARD_URL);
+
+      expect(resp.status).toEqual(200);
+      expect(resp.text).toContain(INTERRUPT_CARD_PAGE_TITLE);
+      expect(resp.text).not.toContain(PAGE_TITLE_ERROR);
+    });
```
Comment (sjahir) [inline test/controllers/update/update.interrupt.card.controller.spec.ts:39]: add an empty line after this

File: test/controllers/update/update.interrupt.card.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    test("catch error when rendering the page", async () => {
+      mockLoggerDebugRequest.mockImplementationOnce( () => { throw new Error(ANY_MESSAGE_ERROR); });
+      const resp = await request(app).get(UPDATE_INTERRUPT_CARD_URL);
+      expect(resp.status).toEqual(500);
+      expect(resp.text).toContain(SERVICE_UNAVAILABLE);
+    });
+    test(`redirect to ${OVERSEAS_ENTITY_QUERY_URL}`, async () => {
```
Comment (sjahir) [inline test/controllers/update/update.interrupt.card.controller.spec.ts:46]: This is a POST request test so it should be in a separate describe block like this:
`describe("POST tests", () => {})`

File: src/controllers/update/update.interrupt.card.controller.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { NextFunction, Request, Response } from "express";
+
+import { logger } from "../../utils/logger";
+import * as config from "../../config";
+
+export const get = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `GET ${config.UPDATE_INTERRUPT_CARD_PAGE}`);
```
Comment (sjahir) [inline src/controllers/update/update.interrupt.card.controller.ts:8]: consider changing this as we have done for our recent tickets (see check.your.answers.controller for an example):
logger.debugRequest(req, `${req.method} ${req.route.path}`);

File: src/controllers/update/update.interrupt.card.controller.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    });
+  } catch (error) {
+    logger.errorRequest(req, error);
+    next(error);
+  }
+};
+
+export const post = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `POST ${config.UPDATE_INTERRUPT_CARD_PAGE}`);
```
Comment (sjahir) [inline src/controllers/update/update.interrupt.card.controller.ts:22]: consider changing this as we have done for our recent tickets (see check.your.answers.controller for an example):
logger.debugRequest(req, `${req.method} ${req.route.path}`);

**Review by sjahir — CHANGES_REQUESTED** [review]

File: test/controllers/update/update.interrupt.card.controller.spec.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
+jest.mock("ioredis");
```
Comment (sjahir) [inline test/controllers/update/update.interrupt.card.controller.spec.ts:1]: Check sonarqube for missing test coverage, just need one more test to cover all the conditions

File: test/controllers/update/update.interrupt.card.controller.spec.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      const resp = await request(app).get(UPDATE_INTERRUPT_CARD_URL);
+
+      expect(resp.status).toEqual(200);
+      expect(resp.text).toContain(INTERRUPT_CARD_PAGE_TITLE);
+      expect(resp.text).not.toContain(PAGE_TITLE_ERROR);
+    });
+
+    test("catch error when rendering the page", async () => {
+      mockLoggerDebugRequest.mockImplementationOnce( () => { throw new Error(ANY_MESSAGE_ERROR); });
+      const resp = await request(app).get(UPDATE_INTERRUPT_CARD_URL);
```
Comment (sjahir) [inline test/controllers/update/update.interrupt.card.controller.spec.ts:43]: add an empty line after this

File: src/controllers/update/update.interrupt.card.controller.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    });
+  } catch (error) {
+    logger.errorRequest(req, error);
+    next(error);
+  }
+};
+
+export const post = (req: Request, res: Response, next: NextFunction) => {
+  try {
+    logger.debugRequest(req, `${req.method} ${req.route.path}`);
```
Comment (sjahir) [inline src/controllers/update/update.interrupt.card.controller.ts:22]: add an empty line after this

**Review by sjahir — CHANGES_REQUESTED** [review]

File: src/routes/index.ts [hunk 13]
```diff
@@ -238,6 +239,11 @@ router.route(config.UPDATE_USE_PAPER_URL)
   .all(authentication)
   .get(updateUsePaper.get);
 
+router.route(config.UPDATE_INTERRUPT_CARD_URL)
```
[the remark this answers is not in the document]
Comment (AndreKramer7654) [reply src/routes/index.ts:242]: I did not see a check in the registration part so not needed.

**Review by sjahir — APPROVED** [review]

**Review by jkayyala — APPROVED** [review]

**Review by bantao-ch — APPROVED** [review]

**Review by mmcpolin-ch — APPROVED** [review]

**Review by mwestacott — APPROVED** [review]

**Review by markpit — APPROVED** [review]

**Pull request merged**